### PR TITLE
Fix generation of temp paths

### PR DESCRIPTION
**Pull request by snowleopard**
The current implementation tries the same candidate 1000 times instead of trying 1000 different candidates. This PR fixes this bug and also does a minor refactoring.

### CODE REVIEW

File: otherlibs/stdune-unstable/temp.ml [hunk 1]
```diff
@@ -58,22 +60,21 @@ let () =
...
 let create ?perms what ~prefix ~suffix =
-  let dir = Filename.get_temp_dir_name () in
+  let dir =
+    Filename.get_temp_dir_name () |> Path.of_filename_relative_to_initial_cwd
+  in
   temp_in_dir ?perms what ~dir ~prefix ~suffix
 
-let temp_in_dir ?perms what ~dir =
-  temp_in_dir ?perms what ~dir:(Path.to_absolute_filename dir)
+let temp_in_dir ?perms what ~dir = temp_in_dir ?perms what ~dir
```
Comment (rgrinberg) [inline otherlibs/stdune-unstable/temp.ml:77]: Do we still need this alias?

Comment (snowleopard) [reply otherlibs/stdune-unstable/temp.ml:77]: Yes, we export this in `temp.mli` and still use in a couple of places.

Comment (rgrinberg) [reply otherlibs/stdune-unstable/temp.ml:77]: What's the point of the alias though? Previously we were doing a path conversion, now we're just doing `let x = x`.

Comment (snowleopard) [reply otherlibs/stdune-unstable/temp.ml:77]: Btw, one of the two use sites, in `src/cache/local.ml`, will soon be gone. Another one is in `dune_engine/dtemp.mli` and I suspect we can get rid of it too.

Comment (snowleopard) [reply otherlibs/stdune-unstable/temp.ml:77]: > What's the point of the alias though? Previously we were doing a path conversion, now we're just doing `let x = x`.

Oh, I got confused. Indeed, we no longer need this alias. Removing.

Comment (snowleopard) [reply otherlibs/stdune-unstable/temp.ml]: Done!

File: otherlibs/stdune-unstable/temp.ml [hunk 2]
```diff
@@ -2,39 +2,41 @@ type what =
   | Dir
   | File
 
-let try_times n ~f =
+let prng = lazy (Random.State.make_self_init ())
+
+let try_paths n ~dir ~prefix ~suffix ~f =
```
Comment (rgrinberg) [inline otherlibs/stdune-unstable/temp.ml:7]: I still don't like this helper very much. In retrospect, this should have been written as:

```
val random_paths : dir:Path.t -> prefix:string -> suffix:string -> Path.t Seq.t
```

And then callers could use it as:

```ocaml
random_paths ~dir ~prefix ~suffix |> Seq.take 1000 |> Seq.find ~f:Path.exists
```

etc.

This isn't pressing however.

**Review by rgrinberg — APPROVED** [review]

File: otherlibs/stdune-unstable/temp.ml (hunk 2 above)
Comment (snowleopard) [reply otherlibs/stdune-unstable/temp.ml:7]: Yes, we can probably improve this in future.

Comment (snowleopard) [reply otherlibs/stdune-unstable/temp.ml:7]: Btw, I'm going to add `with_temp_dir` in another PR, and I'll need this function again. I'll see if I can improve it.

**Comment by ejgallego** [comment]
@snowleopard  , note the 2.8.6. milestone here, so we either need to cherry pick this for 2.9 [and adjust changes], or move the milestone to 3.0

**Comment by snowleopard** [comment]
> @snowleopard , note the 2.8.6. milestone here, so we either need to cherry pick this for 2.9 [and adjust changes], or move the milestone to 3.0

@ejgallego We've actually made quite a few improvements to `temp.ml`, so cherry-picking just one commit seems wrong to me.

Perhaps, we should just leave it to 3.0. Otherwise, instead of looking at individual commits, it may be best to just copy `temp.ml` from `main` and to the 2.9 branch, making necessary adjustments (it should be relatively straightforward).

@rgrinberg Any thoughts on this?

**Comment by ejgallego** [comment]
> Perhaps, we should just leave it to 3.0. Otherwise, instead of looking at individual commits, it may be best to just copy `temp.ml` from `main` to the 2.9 branch, making necessary adjustments (it should be relatively straightforward).

That seems a bit too risky for me at this stage, and I am not knowledgeable in that area as to try that myself.

Thus I propose to target this to 3.0 then [but of course feel free to open a backport PR for 2.9 if you deem the fixes worthy]

**Comment by rgrinberg** [comment]
> Perhaps, we should just leave it to 3.0. Otherwise, instead of looking at individual commits, it may be best to just copy temp.ml from main to the 2.9 branch, making necessary adjustments (it should be relatively straightforward).

I'd say it's fine to push it back until 3.0. In practice, this seems to mainly cause sporadic issues on Windows.

**Comment by snowleopard** [comment]
Pushing back to 3.0 sounds fine!

**Pull request merged**